src/action.rs

```diff
@@ -0,0 +1,165 @@
+use destiny_pkg::TagHash;
+use glam::{Vec2, Vec3};
+
+use crate::camera::FpsCamera;
+use crate::map::{MapList, MapLoadState};
+use crate::overlays::activity_select::{get_activity_hash, set_activity};
+use crate::render::tween::Tween;
+use crate::resources::Resources;
+use std::collections::VecDeque;
+
+pub trait Action {
+    fn start(&mut self, resources: &Resources);
+    fn is_done(&self, resources: &Resources) -> bool;
+    fn is_aborted(&self, resources: &Resources) -> bool;
+}
+
+#[derive(Default)]
+pub struct ActionList {
+    action_queue: VecDeque<Box<dyn Action>>,
+    current_action: Option<Box<dyn Action>>,
+}
+
+impl ActionList {
+    pub fn clear_actions(&mut self) {
+        self.action_queue.clear();
+    }
+
+    pub fn add_action(&mut self, action: impl Action + 'static) {
+        self.action_queue.push_back(Box::new(action));
+    }
+
+    pub fn process(&mut self, resources: &Resources) {
+        let mut clear_current = false;
+        let mut clear_all = false;
+        if let Some(action) = self.current_action.as_ref() {
+            clear_current = action.is_done(resources);
+            clear_all = action.is_aborted(resources);
+        }
+        if clear_current {
+            self.current_action = None;
+        }
+        if clear_all {
+            self.current_action = None;
+            self.action_queue.clear();
+        }
```

Comment:
This angers the borrow checker. I originally wanted to do that. But self.current_action is used in the if. I don't like the way I currently have it, and I'm sure there's some way to make it nicer. I'm just not sure what that would be.

Reply:
We love borrowck 🙃
```suggestion
    if let Some(mut action) = self.current_action.take() {
        if action.is_done(resources) {
            self.current_action = None;
        } else if action.is_aborted(resources) {
            self.current_action = None;
            self.action_queue.clear();
        } else {
            self.current_action = Some(action);
        }
    }
```

Setting current_action to None is of course redundant because we're `take`-ing the value, but i think it looks better semantically